Add support writing large files/empty files

ajeddeloh opened the pull request:
Use io.Reader instead of reading whole files into memory. Prevent
ignition from using large amounts of memory when writing large files.

Also make default http timeout for connection, not the entire transfer

crawford commented on internal/util/http.go:
```diff
@@ -35,7 +36,10 @@ type HttpClient struct {
 	return HttpClient{
 		client: &http.Client{
-			Timeout: 10 * time.Second,
+			//Timeout: 10 * time.Second,
```
Toss this line.

crawford commented on internal/util/http.go:
```diff
@@ -81,6 +85,48 @@ func (c HttpClient) GetWithHeader(url string, header http.Header) ([]byte, int,
 	return body, status, err
 }
 
+//TODO change this comment
```
TODO

crawford commented on internal/util/http.go:
```diff
@@ -81,6 +85,48 @@ func (c HttpClient) GetWithHeader(url string, header http.Header) ([]byte, int,
+
+	err := c.logger.LogOp(func() error {
+		req, err := http.NewRequest("GET", url, nil)
+		req.Header.Set("User-Agent", "Ignition/"+version.Raw)
```
Can we pull these headers into a common location?

crawford commented on internal/util/url.go:
```diff
@@ -107,6 +110,96 @@ func FetchResource(l *log.Logger, u url.URL) ([]byte, error) {
+}
+
+//caller must close reader
+func FetchResourceAsStream(l *log.Logger, u url.URL) (io.ReadCloser, error) {
```
Let's not duplicate so much code.

ajeddeloh commented:
Added HTTP timeouts, docs.

crawford commented on internal/exec/util/file.go:
```diff
@@ -34,61 +35,82 @@ const (
-
-	if l.LogOp(fetch, "fetching file %q", f.Path) != nil {
+	var (
+		contents io.ReadCloser
```
var var var

crawford commented on internal/util/url.go:
```diff
@@ -45,18 +48,51 @@ const (
+	if err := f.Closer.Close(); err != nil {
+		return err
+	}
+	f.umount()
```
always unmount

crawford commented on internal/util/url.go:
```diff
@@ -45,18 +48,51 @@ const (
+	return nil
+}
+
+//caller must close reader
```
clean comment

crawford commented on internal/exec/util/file.go:
```diff
@@ -34,61 +35,82 @@ const (
+		errFunc  func() error
+	)
+
+	contents, err = util.FetchResourceAsStream(l, url.URL(f.Contents.Source))
```
reader

crawford commented on config/types/timeouts.go:
```diff
@@ -0,0 +1,42 @@
+func (t *Timeouts) UnmarshalJSON(data []byte) error {
+	t.HttpResponseHeaders = defaultResponseHeaders
+	t.HttpTotal = defaultTotal
+	th := timeouts(*t)
```
Use a temporary variable for `t` since you are making a copy anyway. No reason to change the Timeouts object.

ajeddeloh replied:
So it doesn't apply defaults if it exits due to error in json.UnmarshalJSON()? In this case don't we want the defaults anyway? Or is there some other reason I'm missing?

crawford replied:
> So it doesn't apply defaults if it exits due to error in json.UnmarshalJSON()?

Correct.

> In this case don't we want the defaults anyway?

No. We shouldn't modify the underlying object unless unmarshalling was successful.

> should the other types that work this way (e.g. filesystems, raid, etc) be changed as well?

Those types work the way I described. They don't need to change.

crawford commented on internal/util/http.go:
```diff
@@ -25,6 +25,17 @@ import (
 
+// A timeout of 0 indicates no timeout
+var (
+	httpResponseHeaderTimeout int
```
I really don't like global variables.

ajeddeloh replied:
What other options are there in this case? The timeouts are applied globally. The only other alternative I see is passing either them (or the Timeouts struct) to every function that might use make http requests, which also doesn't seem very clean

crawford replied:
Pass them as arguments to `NewHttpClient`. If we are calling that too many times, maybe we should reuse HttpClients a bit more.

ajeddeloh replied:
AFAIK we don't reuse them at all and instead create a new one every time we make a request (which is not great). We could switch to something like GetHttpClient, which would hand you either a new one (if not created yet) or one that has already been created, but we would still need a global to store the HttpClient.

crawford replied:
Should probably just reuse them then.

ajeddeloh commented:
Bump, moved the handling of defaults from the config/types to internal/util/http so config can remain a pure representation of the json its passed. Also switched to reusing http clients.

marineam commented on internal/exec/util/file.go:
```diff
@@ -33,62 +36,108 @@ const (
-	decompress := func() error {
-		contents, err = decompressFile(l, f, contents)
-		return err
+	if hasher != nil {
```
This can be nil despite GetHasher reporting no error?

ajeddeloh replied:
Yes, because verification is optional. I wasn't sure whether it would be best to return a nil or something like a ioutil.Discard

marineam commented on internal/exec/util/file.go:
```diff
@@ -33,62 +36,108 @@ const (
+	Mode os.FileMode
+	Uid  int
+	Gid  int
+	hash string
```
Perhaps name this expectedhash or some thing else more descriptive, otherwise it isn't clear what it is compared to the embedded Hash

marineam commented on internal/util/verification.go:
```diff
@@ -47,11 +48,25 @@ func AssertValid(verify types.Verification, data []byte) error {
+
+func GetHasher(verify types.Verification) (hash.Hash, error) {
+	hash := verify.Hash
+	if hash == nil {
```
just use `verify.Hash == nil` and `verify.Hash.Function` below. Not used enough that a local var is helpful and you are masking the package named `hash` which can be confusing.

marineam commented on internal/util/url.go:
```diff
@@ -80,27 +114,40 @@ func FetchResource(l *log.Logger, u url.URL) ([]byte, error) {
+		return struct {
+			io.Reader
+			io.Closer
+		}{bufio.NewReader(f), closer}, nil
```
What is the point of using bufio here? If you are only exposing a ReadCloser you aren't doing anything like pushing input back into the buffer or whatnot.

ajeddeloh replied:
It's reading from a file; don't you usually want to buffer that?

marineam replied:
Usually no

marineam commented on internal/util/url.go:
```diff
@@ -80,27 +114,40 @@ func FetchResource(l *log.Logger, u url.URL) ([]byte, error) {
 		absPath := filepath.Join(oemDirPath, path)
-		data, err := ioutil.ReadFile(absPath)
+		f, err := os.Open(absPath)
+		var closer io.Closer = f
```
why do you need this `closer` variable? Just use f

marineam replied:
oh, I see, you are using a single return. But if you just flip the conditionals you can put `if err == nil { return ... } else if !os.IsExistErr(err) { return err... }` at the top and remove one level of nesting from the bulk of the code.

marineam commented on internal/util/url.go:
```diff
@@ -45,18 +48,50 @@ const (
+	io.Closer
+}
+
+func (f UnmountCloser) Close() error {
```
If you stored the logger instead of a function usage would be a little simpler to follow and close would be:

```go
func (f UnmountCloser) Close() error {
    defer umountOEM(f.logger)
    return f.Closer.Close()
}
```

Or perhaps call it a ReadUnmounter so you implement ReadCloser and avoid wrapping the underlying file in yet another struct below.

marineam commented on internal/exec/util/file.go:
```diff
@@ -33,62 +36,108 @@ const (
-		return err
+	if hasher != nil {
+		hash = f.Contents.Verification.Hash.Sum
+		reader = struct {
```
How about put this in a NewHashedReader function so intent is clear, defining things with anonymous structs inline with real code clutters and distracts a bit.

ajeddeloh commented:
Updated to only have code for conversion to using io.Reader and writing empty files. Switched to using timeout for getting headers instead of whole file when using http. No longer contains code for making timeouts configurable.

crawford commented on internal/exec/util/file.go:
```diff
@@ -33,69 +36,123 @@ const (
+	return nil
+}
+
+//returns a new ReadCloser that also writes to the provided hasher
```
Comments should start with the function name.

crawford commented on internal/exec/util/file.go:
```diff
@@ -33,69 +36,123 @@ const (
-	var contents []byte
+	var reader io.ReadCloser
 	var err error
+	var expectedHash string
```
Why is this needed?

crawford replied:
Nevermind

crawford commented on internal/util/url.go:
```diff
@@ -80,27 +113,46 @@ func FetchResource(l *log.Logger, u url.URL) ([]byte, error) {
+			ReadCloser: f,
+		}, nil
+
+	case "":
```
Break this into a separate commit.

crawford commented on internal/util/url.go:
```diff
@@ -80,27 +113,46 @@ func FetchResource(l *log.Logger, u url.URL) ([]byte, error) {
-			umountOEM(l)
-		} else if err != nil {
+		f, err := os.Open(absPath)
+
```
Remove newline

crawford commented on internal/exec/util/file.go:
```diff
@@ -33,69 +36,123 @@ const (
+	return struct {
+		io.Reader
+		io.Closer
+	}{io.TeeReader(reader, hasher), reader}
```
Use explicit field assignment

crawford commented on internal/exec/util/file.go:
```diff
@@ -33,69 +36,123 @@ const (
+	return &File{
+		Path:         f.Path,
+		ReadCloser:   reader,
+		Hash:         hasher,
```
Can you clean up these names a bit? Assigning `hasher` to `Hash` is odd.

crawford commented on internal/exec/util/file.go:
```diff
@@ -33,69 +36,123 @@ const (
+	Mode         os.FileMode
+	Uid          int
+	Gid          int
+	expectedHash string
```
expectedSum

crawford commented on internal/exec/util/file.go:
```diff
@@ -33,69 +36,123 @@ const (
+}
+
+// Wrapper for gzip's reader that closes the stream it wraps as well
+type gzipReader struct {
```
Is there a better pattern for this in Go? This feels wrong.

ajeddeloh replied:
The problem here is that gzip.NewReader() takes in a io.Reader but returns an io.ReadCloser (interestingly, it implements the io.ReadCloser interface, but the type signature of the function says it returns a io.Reader). Close() on the gzip.Reader does not actually close reader the it is wrapping, which is why I have the extra type that Close()'s both.

crawford replied:
> but the type signature of the function says it returns a io.Reader

Implicit interface implementation is such a great feature.

> Close() on the gzip.Reader does not actually close reader the it is wrapping

Yeah, I noticed. What a pain in the ass.

crawford commented on internal/util/http.go:
```diff
@@ -24,19 +25,49 @@ import (
+)
+
+// Http client to be reused
+var sharedClient *http.Client
```
Still don't like globals.

crawford commented on internal/util/http.go:
```diff
@@ -24,19 +25,49 @@ import (
 )
 
+const (
+	defaultResponseHeadersTimeout int = 10
```
Break this into a different commit.

ajeddeloh commented:
bump with changes sugguested

crawford commented on internal/util/http.go:
```diff
@@ -32,11 +32,13 @@ type HttpClient struct {
 	logger *log.Logger
 }
 
-// NewHttpClient creates a new client with the given logger.
```
Comments should be English sentences. Leave the period.

dghubble commented on internal/util/http.go:
```diff
@@ -36,7 +36,9 @@ type HttpClient struct {
 	return HttpClient{
 		client: &http.Client{
-			Timeout: 10 * time.Second,
+			Transport: &http.Transport{
```
From the net/http package,

> // If nil, DefaultTransport is used.
        Transport RoundTripper

So 524677a replaces the DefaultTransport.

ajeddeloh replied:
Fixed. I though empty value for Transport would be the same as DefaultTransport, but it turns out its not. Copied the default values, expect the ones which don't apply or aren't in go 1.5.

crawford commented:
Please clean up comments according to https://golang.org/doc/effective_go.html#commentary.

crawford commented on internal/exec/util/file.go:
```diff
@@ -109,30 +170,38 @@ func (u Util) WriteFile(f *File) error {
 
 	// Ensure the ownership and mode are as requested (since WriteFile can be affected by sticky bit)
-	if err := os.Chown(tmp.Name(), f.Uid, f.Gid); err != nil {
+	if err = os.Chown(tmp.Name(), f.Uid, f.Gid); err != nil {
```
Break these into a separate commit.

crawford commented on internal/util/url.go:
```diff
@@ -45,18 +47,50 @@ const (
+	var data []byte
+
+	dataReader, err := FetchResourceAsReader(l, u)
+
```
remove line

crawford commented on internal/util/url.go:
```diff
@@ -45,18 +47,50 @@ const (
+	if err != nil {
+		return nil, err
+	}
+
```
remove line

crawford commented on internal/util/url.go:
```diff
@@ -80,27 +113,37 @@ func FetchResource(l *log.Logger, u url.URL) ([]byte, error) {
-			data, err = ioutil.ReadFile(absPath)
-			umountOEM(l)
-		} else if err != nil {
+		f, err := os.Open(absPath)
```
```go
if f, err := os.Open(absPath); err == nil {
    return f, nil
} else if !os.IsNotExist(err) {
```

ajeddeloh commented:
Updated comments, fixed linebreaks, broke error handling fix into seperate commit.

crawford commented:
LGTM. Please rebase.
